COMP: Update jsonxx to latest TransitApp/jsonxx

thewtex opened the pull request:
Update vendored jsonxx from hjiang/jsonxx v1.0.1 (2021) to TransitApp/jsonxx master (eede690, 2025-11-24). TransitApp's fork is the actively maintained continuation with includes cleanup and bug fixes.

<details>
<summary>Changes</summary>

- **jsonxx.h / jsonxx.cc**: Updated to latest TransitApp/jsonxx (eede690). Includes cleanup: `cassert` moved out of header, `string_view` support added, general code improvements.
- **MSVC fix**: Replaced GCC case range extension (`'0'...'9'`) with individual case labels for MSVC compatibility.
- **clang-format**: Applied project clang-format 19.1.7 to vendored files.
- **Rebased** onto current master (v5.4.6 CI, blowekamp's iterator benchmarks, CMake fixes from #76).

</details>

<details>
<summary>Previous review comments</summary>

- jhlegarreta's comment about removing `14` from CMake standard comment — already addressed on master via earlier PRs.

</details>

<!--
provenance: claude-code session 2026-04-14
key_facts: jsonxx upstream is TransitApp/jsonxx (hjiang/jsonxx dormant since 2021); MSVC needs case range fix
related_files: include/jsonxx.h, src/jsonxx.cc
post_merge_action: none
-->

jhlegarreta commented on CMakeLists.txt:
```diff
@@ -1,7 +1,7 @@
 
 if(NOT CMAKE_CXX_STANDARD)
-  set(CMAKE_CXX_STANDARD 14) # Supported values are ``14``, ``17`` and ``20``.
+  set(CMAKE_CXX_STANDARD 17) # Supported values are ``14``, ``17`` and ``20``.
```
Best would be to remove `14` from the comment as well.

thewtex replied:
Good catch!

hjmjohnson replied:
This was addressed on master in an earlier PR — the comment now reads `Supported values are ``17`` and ``20``.` After rebasing onto current master, this is resolved.

jhlegarreta reviewed (commented):
Thanks for doing this Matt.

thewtex commented:
I think we need to move to a different JSON library -- the existing one fails to build with MSVC and is not maintained.

phcerdan commented:
> I think we need to move to a different JSON library -- the existing one fails to build with MSVC and is not maintained.

This one is really famous: https://github.com/nlohmann/json, it seems great to use, but slow performance (not sure about our requirements).

The fastest is simdjson: https://github.com/simdjson/simdjson/#performance-results

thewtex commented:
:Ooo:, simdjson looks nice!

I am also finding glaze:

https://github.com/stephenberry/glaze

Claims to be faster than simdjson, but requires C++20. Has a neat reflection API.

hjmjohnson committed:
COMP: Update jsonxx to latest TransitApp/jsonxx (eede690)

Update vendored jsonxx to TransitApp/jsonxx master (eede690, 2025-11-24)
which includes an includes cleanup removing cassert from the header.
The original hjiang/jsonxx repo (v1.0.1) has not been updated since 2021;
TransitApp fork is the actively maintained continuation.

Replace GCC case range extension with individual case labels for MSVC.

Applied clang-format 19.1.7 to match project style.

hjmjohnson committed:
COMP: Fix MSVC linker errors for jsonxx Empty* globals

Replace `static const Object EmptyObject = {};` (and Array, Value
variants) with `inline` accessor functions returning a local static.
The `static const` form creates per-TU copies whose constructors
reference symbols defined in jsonxx.cc — when the header test TU
links against the shared library on Windows, MSVC cannot resolve
these symbols because they lack __declspec(dllexport).

The `inline` function with a local static is C++17-safe, ODR-correct,
and avoids the linker dependency entirely.

hjmjohnson commented:
Pushed fix for MSVC linker errors (`LNK2019: unresolved external symbol` for `Object::Object`, `Array::Array`, `Value::Value` constructors/destructors).

<details>
<summary>Root cause</summary>

TransitApp/jsonxx declares `static const Object EmptyObject = {};` (and `Array`/`Value` variants) in the header. Each TU that includes the header creates its own copy, calling constructors defined in `jsonxx.cc`. On Windows with shared libraries, these constructors aren't `__declspec(dllexport)`-ed from the DLL, so the header test TU gets `LNK2019`.

Fix: replaced with `inline` accessor functions returning a local static — C++17-safe, ODR-correct, no cross-TU linker dependency.

</details>

hjmjohnson commented:
Windows CI still failing — but this is now an upstream action bug, not a jsonxx issue.

The `v5.4.6` branch of ITKRemoteModuleBuildTestPackageAction received the Windows path-shortening fix (#128 → #129), but that fix introduced a shell compatibility bug: the "Download ITK" step uses `$ITK_BUILD_ROOT` (bash syntax) without `shell: bash`, so on Windows (where GitHub Actions defaults to `pwsh`) the variable is undefined and ITK gets cloned to the wrong location.

Fix PRs: ITKRemoteModuleBuildTestPackageAction#130 (v5.4.6) and #131 (main). Once merged, CI here should be re-triggered.

<details>
<summary>Error trace</summary>

```
CMake Error: The source directory "D:/PerformanceB/ITK" does not exist.
ninja: error: loading 'build.ninja': The system cannot find the file specified.
```

The junction `D:\PerformanceB` → `D:\a\ITKPerformanceBenchmarking` is created correctly, but the Download ITK step's `cd "$ITK_BUILD_ROOT"` is a no-op in PowerShell (needs `$env:ITK_BUILD_ROOT`). ITK gets cloned inside the repo checkout dir instead of at the junction root.

</details>

hjmjohnson committed:
STYLE: Trigger CI with updated action workflow

hjmjohnson reviewed (approved):
Updating to the new vendored code seems to work accross the platforms.

thewtex reviewed (commented):
💚
